go-controller/pkg/cluster/gateway_init.go

```diff
@@ -130,21 +132,35 @@ func (cluster *OvnClusterController) initGateway(
 		}
 	}
 	if err != nil {
-		return nil, nil, err
-	}
-	// marshal the annotation into string
-	bytes, err := json.Marshal(annotations)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to marshal l3 gateway information %v for node %q: (%v)",
-			annotations, nodeName, err)
+		return err
 	}
 
-	nodeAnnotations := map[string]string{ovn.OvnNodeL3GatewayConfig: string(bytes)}
+	if err := nodeAnnotator.Set(ovn.OvnNodeL3GatewayConfig, annotations); err != nil {
+		return err
+	}
 	if systemID != "" {
-		nodeAnnotations[ovn.OvnNodeChassisID] = systemID
+		if err := nodeAnnotator.Set(ovn.OvnNodeChassisID, systemID); err != nil {
+			return err
+		}
 	}
 
-	return nodeAnnotations, prFn, nil
+	// Wait for gateway resources to be created by the master
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := wait.PollImmediate(500*time.Millisecond, 300*time.Second, func() (bool, error) {
+			return gatewayReady()
+		})
+		if err == nil && prFn != nil {
+			// Run the post-ready function
+			err = prFn()
+		}
+		if err != nil {
+			waitErrors <- err
+		}
+	}()
```

Comment:
There's a bunch of duplicated code between `initGateway` and `createManagementPort`, and they have to know how they're expected to use `wg` and `waitErrors` and maybe it would be nice to abstract that out somehow. (`startupReadinessThing.AddWaitFunc(func() error { ... })`)

Reply:
Fixed that up with a startupWaiter object that handles this generically.